main.go

```diff
@@ -1,343 +1,103 @@
-				} else {
-					reports = append(reports, string(stringOut)) //move report to out buffer
-					//lintOut++ (unused)
-				}
-			}
-		}
-		//write to outbuffer
-		outBuffer.Enqueue(reports)
-	}
-	//thread exiting, release hold on main thread
-	poisonBarrier.Done()
-}
-
-// writer thread function for threaded --multi mode
-func writeChunks() {
-	//set-up writer variables
-	var fileWriter *os.File
-	var err error
-	if outPath != "" && outPath != "-" { //only open file if path specified
-		fileWriter, err = os.OpenFile(outPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666) //os.File in write mode
-		if err != nil {                                                                  //file open/write error
-			fmt.Println(err)
-			os.Exit(2) //Fatal Error: Abort
+func ReadCertificate(out chan<- string, filename string, wg *sync.WaitGroup) {
+	log.Info("Reading certificates...")
+	defer wg.Done()
+	if file, err := os.Open(filename); err == nil {
+		defer file.Close()
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			out <- scanner.Text()
 		}
-	}
-
-	for true { //loop until poisoned queue return
-		outStrings := outBuffer.Dequeue() //[]string
-		if outStrings == nil {
-			//queue poioned and empty
-			break //stop writing and prepare for thread exit
-		}
-
-		//print/write all strings in output job
-		for _, x := range outStrings {
-			if outPath != "" && outPath != "-" { //output to file set
-				_, err := fileWriter.WriteString(x)
-				if err != nil { //write failure
-					fmt.Println(err)
-					os.Exit(2) //Fatal Error: Abort
-				}
-				_, err = fileWriter.WriteString("\n")
-				if err != nil { //write failure
-					fmt.Println(err)
-					os.Exit(2) //Fatal Error: Abort
-				}
-			} else { //output to stdout set
-				fmt.Println(x)
-			}
+		if err = scanner.Err(); err != nil {
+			log.Fatal("Error with scanning file: ", err)
 		}
+	} else {
+		log.Fatal("Error reading file: ", err)
 	}
-	//writeing complete, clean-up and release main thread to exit
-	fileWriter.Close()
-
-	//entering main thread exit barrier
-	exittex.Lock()
-
-	writeComplete = true
-	mainWait.Signal() //wake main thread in case it beat us here
-
-	//exiting main thread exit barrier
-	exittex.Unlock()
 }
 
-func singleMode() ([]byte, error) {
-
-	theCert := lints.ReadCertificate(inPath)
+func main() {
+	log.SetLevel(log.InfoLevel)
+	runtime.GOMAXPROCS(numProcs)
 
-	if theCert == nil {
-		return nil, errors.New("Parsing Failed")
-	}
-	theReport, err := zlint.ParsedTestHandler(theCert)
-	if err != nil {
-		return nil, err
+	if prettyPrint {
+		zlint.PrettyPrintZLint()
+		return
 	}
 
-	return json.Marshal(theReport)
+	//Initialize Channel
+	certs := make(chan string, channelSize)
+	var readerWG sync.WaitGroup
+	var procWG sync.WaitGroup
+	readerWG.Add(1)
+	go ReadCertificate(certs, inPath, &readerWG)
+	for i := 0; i < numCertThreads; i++ {
+		procWG.Add(1)
+		go ProcessCertificate(certs, &procWG)
+	}
+	go func() {
+		readerWG.Wait()
+		close(certs)
+	}()
+	procWG.Wait()
```

Comment:
I'd change concurrency setup here a little to something like:

1 input thread reads and splits input -> N threads that parse and run zlint and encode to JSON -> 1 thread that writes output.

A case could also be made for:

1 input thread that splits input -> N threads that take sections of input and decode to DER bytes -> N threads that parse certificate and run zlint -> N threads that take cert + zlint and encode to JSON -> 1 thread that writes output.

(Or some combination of that breakdown, I could see a couple setups).

At the end of the day, I'd still want to see a real clear high level main. You're main is good now, but you shouldn't be outputting from each worker goroutine.

Additionally, you should check out some of the code I hacked together in zcertificate for handling different input formats, including the `ScannerFunc` I wrote to split a file by PEM. It might not handle text between PEM's though. We'll want to handle both a PEM file with multiple certificates, and a "csv" of one base64-encoded DER certificate per line.

Overall, WAYYY better than it was previously.